lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +433,811 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+            auto *I = Succ->getFirstNonPHI();
+
+            if (isa<CallInst>(I)
+                && cast<CallInst>(I)->getCalledFunction() == PreHookMarker) {
+              MetaAddress Destination;
+              model::FunctionEdgeType::Values Type;
+              auto *CalleePC = cast<CallInst>(I)->getArgOperand(1);
+
+              // Direct or indirect call?
+              if (isa<ConstantStruct>(CalleePC)) {
+                Destination = MetaAddress::fromConstant(CalleePC);
+                Type = model::FunctionEdgeType::FunctionCall;
+              } else {
+                Destination = MetaAddress::invalid();
+                Type = model::FunctionEdgeType::IndirectCall;
+              }
+
+              auto Edge = makeEdge(Destination, Type);
+              auto DestTy = Oracle.getFunctionType(Destination);
+              if (DestTy == FunctionTypeValue::NoReturn)
+                cast<model::CallEdge>(Edge.get())->IsNoReturn = true;
+
+              Block.Successors.insert(Edge);
+            } else if (auto *Call = getCallTo(I, "function_call")) {
+              // Marker can be erased now, no longer needed.
+              Call->eraseFromParent();
+
+              auto Destination = getBasicBlockPC(Succ);
+              auto Edge = makeEdge(Destination,
+                                   model::FunctionEdgeType::FakeFunctionCall);
+              Block.Successors.insert(Edge);
+            } else {
+              Queue.insert(Succ);
+            }
+          }
+        }
+      }
+
+      Block.End = End + Size;
+      CFG.insert(Block);
+    }
+  }
+
+  return CFG;
+}
+
+template<class FO>
+FunctionSummary
+CFEPAnalyzer<FO>::analyze(GlobalsArrayRef ABIRegisters,
+                          llvm::BasicBlock *Entry,
+                          llvm::raw_fd_ostream &OutputIBI,
+                          llvm::raw_fd_ostream &OutputAAWriter) {
+  using namespace llvm;
+
+  Value *RA = nullptr;
+  const auto *PCH = GCBI->programCounterHandler();
+  StructType *MetaAddressTy = MetaAddress::getStruct(&M);
+
+  // Run function boundaries detection, and retrieve the outlined function.
+  Function *OutFunc = createDisposableFunction(Entry);
+
+  // Initialize CFG of the function for the model.
+  auto CFG = collectDirectCFG(Entry, OutFunc);
+
+  // Identify the callee-saved registers and tell if a function returns
+  // properly, i.e., it jumps to the return address. To achieve this, we craft
+  // the IR by loading the PC, the SP, as well as the ABI registers CSVs
+  // respectively at function entry and end. We win when the subtraction between
+  // their entry and end values is zero.
+  IRBuilder<> Builder(&OutFunc->getEntryBlock().front());
+  auto *IntPtrTy = GCBI->spReg()->getType();
+  auto *IntTy = GCBI->spReg()->getType()->getElementType();
+
+  auto *SP = Builder.CreateLoad(GCBI->spReg());
+  auto *SPPtr = Builder.CreateIntToPtr(SP, IntPtrTy);
+  auto *GEPI = Builder.CreateGEP(IntTy, SPPtr, ConstantInt::get(IntTy, 0));
```

Comment:
This is suspicious.
In 64-bit architecture, `GCBI->spReg()` is a `uint64_t *`. You load it and you get a `uint64_t`, but then you `IntToPtr` it into a `uint64_t *`.
What's the ratio behind this?
Are we assuming the stack is a set of register-sized slots?
If we were to switch to `uint8_t *`, what would happen?

Reply:
This is because we need to take the address of the load of `SP` and pass it to the pointer type of the `getelementptr`, to perform address subtraction between the initial GEP and final one. If we were to switch to `uint8_t *`, I don't see why it should not work. LLVM optimizations oftentimes transform the type of the GEP to `i8`, if any.

Reply:
So, there's not benefit in using `uint64_t *` instead of `uint8_t *`?
If so, let's make the most obvious choice, which is `uint8_t *`.
Unless there's a benefit, I prefer not to have a situation where we have `uint64_t *` on 64-bit architectures and `uint32_t *` on 32-bit architectures.

Reply:
Sorry, but do we really want this? Isn't it better to leverage on the type of `GCBI->spReg` instead of crafting a dedicated `i8` for this? Will `uint8_t *` work for any architecture we (will) support?

Reply:
"Leverage" implies a benefit. What is the benefit here?
If there's no benefit, I prefer things to be coherent across architectures.

Reply:
I don't see any reason why this should not work, but do you?

Reply:
I vaguely remember at the beginning we didn't decide to directly use a `uint8_t *`, is that possible?

Reply:
Why?

Reply:
We need to understand why `uint64_t *` works but `uint8_t *` doesn't, we can't have parts of the code we can't explain why are there.

Let me know if you need any help with this.

Reply:
Alright, I'll be going deeper on this.